File: Algorithms/list5/C++/task3/Binheap.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <cassert>

struct BinomialNode {
    int key;
    int degree;
    BinomialNode* parent;
    BinomialNode* child;
    BinomialNode* sibling;

    BinomialNode(int k) : key(k), degree(0), parent(nullptr), child(nullptr), sibling(nullptr) {}
};

class BinomialHeap {
private:
    BinomialNode* head;
    // Total comparisons for all ops and current op count
    long long total_comparisons = 0;
    long long current_op_comparisons = 0;

    // Compare keys and count comparisons
    bool less_or_equal(int a, int b) {
        current_op_comparisons++;
        return a <= b;
    }

    // Link two binomial trees of the same degree
    void linkTrees(BinomialNode* y, BinomialNode* z) {
        y->parent = z;
        y->sibling = z->child;
        z->child = y;
        z->degree++;
    }

    // Merge root lists of two heaps by degree ascending
    BinomialNode* mergeRootLists(BinomialNode* h1, BinomialNode* h2) {
        if (!h1) return h2;
        if (!h2) return h1;

        BinomialNode* head = nullptr;
        BinomialNode** pos = &head;

        while (h1 && h2) {
            if (h1->degree < h2->degree) {
                *pos = h1;
                h1 = h1->sibling;
            } else {
                *pos = h2;
                h2 = h2->sibling;
            }
            pos = &((*pos)->sibling);
        }

        *pos = (h1) ? h1 : h2;
        return head;
    }

    // Union of two heaps
    BinomialNode* unionHeaps(BinomialNode* h1, BinomialNode* h2) {
        BinomialNode* newHead = mergeRootLists(h1, h2);
        if (!newHead) return nullptr;

        BinomialNode* prev = nullptr;
        BinomialNode* curr = newHead;
        BinomialNode* next = curr->sibling;

        while (next) {
            if ( (curr->degree != next->degree) ||
                 (next->sibling && next->sibling->degree == curr->degree)) {
                prev = curr;
                curr = next;
            } else {
                // Compare keys to link
                current_op_comparisons++;
                if (curr->key <= next->key) {
                    curr->sibling = next->sibling;
                    linkTrees(next, curr);
                } else {
                    if (prev) prev->sibling = next;
                    else newHead = next;
                    linkTrees(curr, next);
                    curr = next;
                }
            }
            next = curr->sibling;
        }
        return newHead;
    }

public:
    BinomialHeap() : head(nullptr) {}

    void resetComparisonCount() {
        current_op_comparisons = 0;
    }

    void addToTotalComparisons() {
        total_comparisons += current_op_comparisons;
    }

    long long getCurrentOpComparisons() const {
        return current_op_comparisons;
    }

    long long getTotalComparisons() const {
        return total_comparisons;
    }

    void clear() {
        // TODO: Implement recursive destructor for nodes if needed
        head = nullptr;
        total_comparisons = 0;
        current_op_comparisons = 0;
    }

    void makeHeap() {
        head = nullptr;
        total_comparisons = 0;
        current_op_comparisons = 0;
    }

    void insert(int key) {
        resetComparisonCount();
        BinomialNode* newNode = new BinomialNode(key);
        head = unionHeaps(head, newNode);
        addToTotalComparisons();
    }

    void unionHeap(BinomialHeap& other) {
        resetComparisonCount();
        head = unionHeaps(head, other.head);
        other.head = nullptr; // empty the other heap
        addToTotalComparisons();
    }

    bool empty() const {
        return head == nullptr;
    }

    // Extract-Min returns min key or -1 if empty
    int extractMin() {
        resetComparisonCount();
        if (!head) return -1;

        // Find min root and keep track of prev
        BinomialNode* minNode = head;
        BinomialNode* minPrev = nullptr;

        BinomialNode* curr = head;
        BinomialNode* prev = nullptr;

        while (curr) {
            current_op_comparisons++;
            if (curr->key < minNode->key) {
                minNode = curr;
                minPrev = prev;
            }
            prev = curr;
            curr = curr->sibling;
        }

        // Remove minNode from root list
        if (minPrev) {
            minPrev->sibling = minNode->sibling;
        } else {
            head = minNode->sibling;
        }

        // Reverse minNode's children list to form new heap
        BinomialNode* child = minNode->child;
        BinomialNode* newHead = nullptr;
        while (child) {
            BinomialNode* next = child->sibling;
            child->sibling = newHead;
            child->parent = nullptr;
            newHead = child;
            child = next;
        }

        head = unionHeaps(head, newHead);

        int minKey = minNode->key;
        delete minNode;

        addToTotalComparisons();
        return minKey;
    }
};

// Helper: generate vector of random ints size n
std::vector<int> randomSequence(int n, int seed) {
    std::mt19937 rng(seed);
    std::uniform_int_distribution<int> dist(1, 1000000);
    std::vector<int> v(n);
    for (int &x : v) x = dist(rng);
    return v;
}

// Experiment for n=500, 5 trials, printing comparison counts for each op
void runExperimentN500() {
    const int n = 500;
    for (int trial = 1; trial <= 20; trial++) {
        std::cout << "Trial " << trial << ":\n";
        BinomialHeap H1, H2;

        H1.makeHeap();
        H2.makeHeap();

        auto seq1 = randomSequence(n, trial * 100);
        auto seq2 = randomSequence(n, trial * 200);

        std::vector<long long> insert_comps_H1;
        std::vector<long long> insert_comps_H2;

        // Insert into H1
        for (int x : seq1) {
            H1.insert(x);
            insert_comps_H1.push_back(H1.getCurrentOpComparisons());
        }

        // Insert into H2
        for (int x : seq2) {
            H2.insert(x);
            insert_comps_H2.push_back(H2.getCurrentOpComparisons());
        }

        // Union
        H1.resetComparisonCount();
        H1.unionHeap(H2);
        long long union_comps = H1.getCurrentOpComparisons();

        // Extract min 2n times
        std::vector<long long> extract_comps;
        std::vector<int> extracted;

        bool sorted = true;
        int prev = -1;

        for (int i = 0; i < 2 * n; i++) {
            int val = H1.extractMin();
            extracted.push_back(val);
            extract_comps.push_back(H1.getCurrentOpComparisons());

            if (i > 0 && extracted[i] < extracted[i - 1]) sorted = false;
        }

        // Check empty heap after extractions
        bool empty_after = H1.empty();

        // Output summary
        std::cout << "Insert H1 comparisons (per insert): ";
        for (auto c : insert_comps_H1) std::cout << c << " ";
        std::cout << "\nInsert H2 comparisons (per insert): ";
        for (auto c : insert_comps_H2) std::cout << c << " ";
        std::cout << "\nUnion comparisons: " << union_comps << "\n";
        std::cout << "Extract-Min comparisons (per op): ";
        for (auto c : extract_comps) std::cout << c << " ";
        std::cout << "\nExtracted sequence sorted: " << (sorted ? "YES" : "NO") << "\n";
        std::cout << "Heap empty after 2n extracts: " << (empty_after ? "YES" : "NO") << "\n";
        std::cout << "---------------------------------------\n";
    }
}

// Experiment scaling n with output of total comparisons / n (average cost per operation)
void runScalingExperiments() {
    std::cout << "N ; AVG_comparisons_per_operation\n";
    for (int n = 100; n <= 10000; n += 100) {
        BinomialHeap H1, H2;

        H1.makeHeap();
        H2.makeHeap();

        auto seq1 = randomSequence(n, n * 1000 + 1);
        auto seq2 = randomSequence(n, n * 2000 + 2);

        for (int x : seq1) H1.insert(x);
        for (int x : seq2) H2.insert(x);

        H1.unionHeap(H2);

        for (int i = 0; i < 2 * n; i++) H1.extractMin();

        // Average comparisons per operation
        long long total_ops = 2 * n + 2 * n + 1; // approx: inserts + extracts + union
        // Actually total ops: insertions (2*n), union(1), extractions(2*n) = 4n+1

        long long total_comparisons = H1.getTotalComparisons();

        double avg_comp = double(total_comparisons) / (4.0 * n + 1);

        std::cout << n << " ; " << avg_comp << "\n";
    }
}

int main() {
    // Run experiment for n=500 (5 trials)
    runExperimentN500();

    // Run scaling experiments
    runScalingExperiments();

    return 0;
}

File: Algorithms/list5/Haskell/task3/BinHeap.hs
module BinHeap where

import Data.IORef
import System.IO.Unsafe (unsafePerformIO)

-- Define a binomial tree node: rank, value, list of child trees
data BinomialTree a = Node Int a [BinomialTree a] deriving Show

-- Wrap the heap in a newtype for better encapsulation
newtype BinomialHeap a = BinomialHeap [BinomialTree a] deriving Show

-- Get the rank (degree) of a tree
rank :: BinomialTree a -> Int
rank (Node r _ _) = r

-- Get the root value of a tree
root :: BinomialTree a -> a
root (Node _ x _) = x

-- Get the child trees of a node
children :: BinomialTree a -> [BinomialTree a]
children (Node _ _ c) = c

-- Link two trees of the same rank: the smaller root becomes parent
link :: Ord a => BinomialTree a -> BinomialTree a -> BinomialTree a
link t1@(Node r x1 c1) t2@(Node _ x2 c2)
    | x1 <= x2  = Node (r + 1) x1 (t2 : c1)  -- t2 becomes child of t1
    | otherwise = Node (r + 1) x2 (t1 : c2)  -- t1 becomes child of t2

-- Insert a value into the heap
insert :: Ord a => a -> BinomialHeap a -> BinomialHeap a
insert x (BinomialHeap ts) = BinomialHeap (insertTree (Node 0 x []) ts)

-- Insert a tree into a heap, ensuring no duplicate ranks
insertTree :: Ord a => BinomialTree a -> [BinomialTree a] -> [BinomialTree a]
insertTree t [] = [t]
insertTree t (t':ts)
    | rank t < rank t' = t : t' : ts
    | otherwise        = insertTree (link t t') ts  -- combine same-rank trees

-- Merge two heaps
merge :: Ord a => BinomialHeap a -> BinomialHeap a -> BinomialHeap a
merge (BinomialHeap h1) (BinomialHeap h2) = BinomialHeap (mergeTrees h1 h2)

-- Merge the underlying tree lists
mergeTrees :: Ord a => [BinomialTree a] -> [BinomialTree a] -> [BinomialTree a]
mergeTrees [] h = h
mergeTrees h [] = h
mergeTrees h1@(t1:ts1) h2@(t2:ts2)
    | rank t1 < rank t2 = t1 : mergeTrees ts1 h2
    | rank t2 < rank t1 = t2 : mergeTrees h1 ts2
    | otherwise         = insertTree (link t1 t2) (mergeTrees ts1 ts2)

-- Find and remove the tree with the minimum root value
removeMinTree :: Ord a => [BinomialTree a] -> (BinomialTree a, [BinomialTree a])
removeMinTree [] = error "empty"
removeMinTree [t] = (t, [])
removeMinTree (t:ts) =
    let (t', ts') = removeMinTree ts
    in if root t <= root t' then (t, ts) else (t', t : ts')

-- Extract the minimum value and return the new heap without it
extractMin :: Ord a => BinomialHeap a -> (a, BinomialHeap a)
extractMin (BinomialHeap ts) =
    let (Node _ x c, rest) = removeMinTree ts
        newHeap = merge (BinomialHeap (reverse c)) (BinomialHeap rest)
    in (x, newHeap)

-- COMPARISON COUNTING FUNCTIONS (renamed to avoid clash)

{-# NOINLINE comparisonCount #-}
comparisonCount :: IORef Int
comparisonCount = unsafePerformIO (newIORef 0)

resetComparisonCount :: IO ()
resetComparisonCount = writeIORef comparisonCount 0

incComparisonCount :: IO ()
incComparisonCount = modifyIORef' comparisonCount (+1)

getComparisonCount :: IO Int
getComparisonCount = readIORef comparisonCount
